infrastructure/smart-contracts/test/factory/AchievementsFactoryTests.mjs

```diff
@@ -0,0 +1,348 @@
+
+            //assign MINT_ROLE to new account
+            const newMinterAddress = await addr2.getAddress();
+            await achievementsFactory.connect(addr1).grantRole(mintRoleHash, newMinterAddress);
+            
+            //mint tokens as new minter account
+            const toAddress = await addr1.getAddress();
+            const tokenId = 0;
+            const data = "0x";
+            let trx2 = await tokenContract.connect(addr2).mint(toAddress, tokenId, data);
+            let rec2 = await trx2.wait();
+
+            //assert contract state and events
+            expect(rec2.logs[0].fragment.name).to.equal("TransferSingle");
+            expect(rec2.logs[0].args[0]).to.equal(newMinterAddress);
+            expect(rec2.logs[0].args[1]).to.equal(ethers.ZeroAddress);
+            expect(rec2.logs[0].args[2]).to.equal(toAddress);
+            expect(rec2.logs[0].args[3]).to.equal(tokenId);
+            expect(rec2.logs[0].args[4]).to.equal(1);
+            expect(await tokenContract.balanceOf(toAddress, tokenId)).to.equal(1);
+        });
+
+        it("should fail to mint more than one token of token id to account", async function() {
+            const {
+                achievementsFactory, 
+                addr1
+            } = await loadFixture(deployInfrastructure);
+
+            //produce new token contract
+            const gameId = "test-game-id";
+            const trx = await achievementsFactory.connect(addr1).produceContract(gameId, baseURI);
+            const rec = await trx.wait();
+            const tokenContractAddress = rec.logs[0].args[0];
+            const AchievementsContractFactory = await ethers.getContractFactory("Achievements");
+            const tokenContract = await AchievementsContractFactory.attach(tokenContractAddress);
+            
+            //mint initial token
+            const toAddress = await addr1.getAddress();
+            const tokenId = 0;
+            const data = "0x";
+            await tokenContract.connect(addr1).mint(toAddress, tokenId, data);
+
+            //attempt to mint another token with same tokenid to same address
+            await expect(
+				tokenContract.connect(addr1).mint(toAddress, tokenId, data)
+			).to.be.revertedWith("address has non-zero token balance");
+
+            //attempt to batch mint another token with same tokenid to same address
+            const tokenIds = [tokenId];
+            await expect(
+				tokenContract.connect(addr1).mintBatch(toAddress, tokenIds, data)
+			).to.be.revertedWith("address has non-zero token balance");
+
+            //attempt to batch mint multiple tokens with same tokenid
+            const unmintedTokenId = 1;
+            const duplicateTokenIds = [unmintedTokenId, unmintedTokenId];
+            await expect(
+				tokenContract.connect(addr1).mintBatch(toAddress, duplicateTokenIds, data)
+			).to.be.revertedWith("address has non-zero token balance");
```

Comment:
Will this work for one id that is already minted and a second one that is not ?
This could help us a lot when syncing.

Reply:
This would cause a revert in the event that the ids array is invalid.